### PR TITLE
Update i18n middleware and view to use locale_id in user profile

### PR DESCRIPTION
**Pull request by giorgio93p**
This pull request incorporates the addition of `locale_id` to `UserProfile` into i18n middleware and locale changing view.

It is not ready yet: integration tests fail because the users added in integration test helpers do not have a profile. @adamhooper, I would like some advice on it.

### CODE REVIEW

**Comment by adamhooper** [comment]
D'oh -- that's an error in the integration-test suite. I'll update the test suite very soon; then you can rebase against master.

We should actually use integration tests to test language selection, in `integrationtests/test_I18n.py`. But I'm okay with deploying first and integration-testing later.

**Comment by giorgio93p** [comment]
Thank you

I have already added a todo in my personal list for an integration test for locale switching. I haven't done it because we didn't have a final (as final it can be!) design for the switcher.

**Comment by adamhooper** [comment]
@giorgio93p I just pushed commit 98bc4c8c6598f390a544b3beb58bcf106331d6e1 which adds a UserProfile for every user in integration tests. (We should already assume every user has a UserProfile on production.)

**Comment by giorgio93p** [comment]
Although in principle not related to this pull request, I also pushed a commit that shows each locale in its own language.

**Comment by giorgio93p** [comment]
Integration test passes on my machine; I don't have access rights to see where it falied in Google Container Builder. It's not the first time this happens.

**Comment by adamhooper** [comment]
`bin/dev unittest`

File: cjworkbench/middleware/i18n.py [hunk 1]
```diff
@@ -64,11 +64,11 @@ def _get_locale_from_current_user(self, request):
             locale = request.session.get("locale_id")
             return locale if is_supported(locale) else None
         else:
-            locale = getattr(request.user, "locale_id", None)
+            locale = getattr(request.user.user_profile, "locale_id", None)
             return locale if is_supported(locale) else None
 
     def _use_session(self, user):
-        return not user.is_authenticated or not hasattr(user, "locale_id")
+        return not user.is_authenticated or not hasattr(user.user_profile, "locale_id")
```
Comment (adamhooper) [inline cjworkbench/middleware/i18n.py]: When does `user_profile` not have a `locale_id` attribute?

File: cjworkbench/i18n/views.py [hunk 2]
```diff
@@ -23,4 +23,7 @@ def set_locale(request):
     locale = request.POST.get("new_locale")
     if is_supported(locale):
         request.session["locale_id"] = locale
+        if request.user.is_authenticated:
+            request.user.user_profile.locale_id = locale
+            request.user.user_profile.save()
```
Comment (adamhooper) [inline cjworkbench/i18n/views.py]: Never call `Model.save()` without arguments. It's race-prone.

Use `request.user.user_profile.save(update_fields=['locale_id'])`

(Django ORM slogan: "make easy tasks hard and hard tasks impossible.")

File: cjworkbench/tests/test_i18n_middleware.py [hunk 3]
```diff
@@ -15,11 +15,15 @@ def mock_response(request):
     return response
 
 
+class MockUserProfile:
+    def __init__(self, locale_preference=None):
+        self.locale_id = locale_preference
+
+
 class MockUser:
     def __init__(self, locale_preference=None):
```
Comment (adamhooper) [inline cjworkbench/tests/test_i18n_middleware.py]: Call the param `user_profile__locale_id`

File: cjworkbench/tests/test_i18n_middleware.py [hunk 4]
```diff
@@ -15,11 +15,15 @@ def mock_response(request):
     return response
 
 
+class MockUserProfile:
+    def __init__(self, locale_preference=None):
```
Comment (adamhooper) [inline cjworkbench/tests/test_i18n_middleware.py]: Call this param `locale_id`.

**Comment by adamhooper** [comment]
@giorgio93p Since it's Friday night, I won't merge this. But you're welcome to merge it Monday morning before I get in.

**Pull request merged**